src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1803,188 +1803,270 @@ public static bool Decompose(Matrix4x4 matrix, out Vector3 scale, out Quaternion
-                    CanonicalBasis canonicalBasis = default;
-                    Vector3* pCanonicalBasis = &canonicalBasis.Row0;
+                    pVectorBasis_A = pVectorBasis_2;
+                    pVectorBasis_B = pVectorBasis_1;
+                    pVectorBasis_C = pVectorBasis_0;
 
-                    canonicalBasis.Row0 = new Vector3(1.0f, 0.0f, 0.0f);
-                    canonicalBasis.Row1 = new Vector3(0.0f, 1.0f, 0.0f);
-                    canonicalBasis.Row2 = new Vector3(0.0f, 0.0f, 1.0f);
+                    canonicalBasis_A = canonicalBasis_Row2;
+                }
+                else
+                {
+                    a = 1;
+                    pfScales_A = pfScales_1;
+                    pVectorBasis_A = pVectorBasis_1;
+                    canonicalBasis_A = canonicalBasis_Row1;
+
+                    if (x < z)
+                    {
+                        b = 2;
 
-                    translation = new Vector3(
-                        matrix.M41,
-                        matrix.M42,
-                        matrix.M43);
+                        pfScales_B = pfScales_2;
+                        pfScales_C = pfScales_0;
 
-                    pVectorBasis[0] = (Vector3*)&matTemp.M11;
-                    pVectorBasis[1] = (Vector3*)&matTemp.M21;
-                    pVectorBasis[2] = (Vector3*)&matTemp.M31;
+                        pVectorBasis_B = pVectorBasis_2;
+                        pVectorBasis_C = pVectorBasis_0;
+                    }
+                    else
+                    {
+                        b = 0;
 
-                    *(pVectorBasis[0]) = new Vector3(matrix.M11, matrix.M12, matrix.M13);
-                    *(pVectorBasis[1]) = new Vector3(matrix.M21, matrix.M22, matrix.M23);
-                    *(pVectorBasis[2]) = new Vector3(matrix.M31, matrix.M32, matrix.M33);
+                        pfScales_B = pfScales_0;
+                        pfScales_C = pfScales_2;
 
-                    scale.X = pVectorBasis[0]->Length();
-                    scale.Y = pVectorBasis[1]->Length();
-                    scale.Z = pVectorBasis[2]->Length();
+                        pVectorBasis_B = pVectorBasis_0;
+                        pVectorBasis_C = pVectorBasis_2;
+                    }
+                }
+            }
+            else
+            {
+                if (x < z)
+                {
+                    a = 2;
+                    b = 0;
 
-                    uint a, b, c;
-                    #region Ranking
-                    float x = pfScales[0], y = pfScales[1], z = pfScales[2];
-                    if (x < y)
+                    pfScales_A = pfScales_2;
+                    pfScales_B = pfScales_0;
+                    pfScales_C = pfScales_1;
+
+                    pVectorBasis_A = pVectorBasis_2;
+                    pVectorBasis_B = pVectorBasis_0;
+                    pVectorBasis_C = pVectorBasis_1;
+
+                    canonicalBasis_A = canonicalBasis_Row2;
```

Comment:
This logic could be simplified by just using `a`, `b`, and `c` as indices instead